### PR TITLE
DOC: Minor enhancements to documenting_mpl

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

This is a follow-on from #9805

Includes anntzer's suggested changes for the artist introspection phrase...

Includes my suggestions for a bit more of an overview at the beginning of what comes from where in the build process.

I think this document could still be improved, but that can be longer-term.  I thought I'd submit these changes while they are on everyone's minds...  See outline at #9876

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is PEP 8 compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: doc/devel/documenting_mpl.rst [hunk 1]
```diff
@@ -136,28 +144,54 @@ An example docstring looks like:
...
 
   `~matplotlib.collections.LineCollection`
 
-It is also possible to add links to code in Python, Numpy, Scipy, or Pandas.
-Sometimes it is tricky to get external Sphinx linking to work; to check that
-a something exists to link to the following shell command outputs a list of all
-objects that can be referenced (in this case for Numpy)::
+returns a link to the documentation of
+`~matplotlib.collections.LineCollection`.  If we want the full path of the
+class to be shown then we omit the tilde:
```
Comment (dstansby) [inline doc/devel/documenting_mpl.rst]: 😮 I didn't know this!

Comment (jklymak) [reply doc/devel/documenting_mpl.rst]: Yeah, @anntzer explained it all to me (somewhere).  Its also nicely explained at the very bottom of this very long section in the sphinx docs:  

http://www.sphinx-doc.org/en/stable/domains.html#the-python-domain

Could the discoverability of this be better?  Probably!

**Comment by jklymak** [comment]
For review:   https://4050-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/devel/documenting_mpl.html

File: doc/devel/documenting_mpl.rst [hunk 2]
```diff
@@ -136,28 +144,54 @@ An example docstring looks like:
...
+
+.. code-block:: python
+
+  `matplotlib.collections.LineCollection`
+
+to get `matplotlib.collections.LineCollection`.  Its often not
+necessary to fully specify the class hierarchy:
```
Comment (anntzer) [inline doc/devel/documenting_mpl.rst]: often not = unless there's a collision

File: doc/devel/documenting_mpl.rst [hunk 3]
```diff
@@ -43,16 +43,17 @@ guide, developers guide, api reference, and FAQs. The documentation suite is
...
-   and :file:docs/tutorials, as they are rebuilt from the original sources in
-   the root directory.
-
+Sphinx_ also creates ``.rst`` files that are staged in :file:`doc/api` from
+the docstrings of the classes in the Matplotlib library.  Except for
+:file:`doc/api/api_changes/`, these ``.rst`` are created when the
+documentation is built.
+
+Similarly, the contents of :file:`docs/gallery` and :file:`docs/tutorials` are
+generated by the `Sphinx Gallery`_ from the sources in :file:`examples` and
```
Comment (anntzer) [inline doc/devel/documenting_mpl.rst]: not directly related but I think the "examples" folder should be renamed "gallery" or the "gallery" folder on the website should be renamed "examples"...

File: doc/devel/documenting_mpl.rst [hunk 4]
```diff
@@ -136,28 +144,54 @@ An example docstring looks like:
...
+links just as well: `~.LineCollection`.
+
+Other packages can also be linked via ``intersphinx``:
+
+.. code-block:: Python
+
+  `numpy.mean`
+
+will return this link: `numpy.mean`.  This works for Python, Numpy, Scipy,
+and Pandas. Sometimes it is tricky to get external Sphinx linking to work; to
```
Comment (anntzer) [inline doc/devel/documenting_mpl.rst]: note that the list of packages for which it works is in conf.py

File: doc/devel/documenting_mpl.rst [hunk 5]
```diff
@@ -136,28 +144,54 @@ An example docstring looks like:
...
+check that a something exists to link to the following shell command outputs
+a list of all objects that can be referenced (in this case for Numpy)::
 
   python -m sphinx.ext.intersphinx 'https://docs.scipy.org/doc/numpy/objects.inv'
 
 
 Function arguments
-------------------
+~~~~~~~~~~~~~~~~~~
+
```
Comment (anntzer) [inline doc/devel/documenting_mpl.rst]: may be worth explaining briefly why backquotes doesn't work (arguments don't have an entry)

File: doc/devel/documenting_mpl.rst (hunk 3 above)
Comment (jklymak) [reply doc/devel/documenting_mpl.rst]: Yeah I agree.  Its very confusing.  I prefer "examples".

File: doc/devel/documenting_mpl.rst (hunk 2 above)
Comment (jklymak) [reply doc/devel/documenting_mpl.rst]: Added `unless there is a namespace collision between two packages`

File: doc/devel/documenting_mpl.rst (hunk 5 above)
Comment (jklymak) [reply doc/devel/documenting_mpl.rst]: Changed to:

```rst

   Do not describe `argument` like this.  As per the previous section,
   this syntax will attempt to resolve the argument as a link to an object in the
   library.
```

Comment (anntzer) [reply doc/devel/documenting_mpl.rst]: will (unsuccessfully) attempt to etc.

**Comment by jklymak** [comment]
Rebased off master, so it includes changes in #9875 

https://circleci.com/gh/matplotlib/matplotlib/4091/artifacts/0//home/circleci/project/doc/build/html/index.html

**Comment by jklymak** [comment]
#9884 includes these changes and more, so closing this in favour...

**Comment by jklymak** [comment]
Can re-open if #9884 doesn't go anywhere....

**Pull request closed without merging**